src/dhtproto/client/internal/NodeHashRanges.d

```diff
@@ -57,6 +59,115 @@ public final class NodeHashRanges : NodeHashRangesBase
 
     /***************************************************************************
 
+        Helper encapsulating the node-selection logic required by requests that
+        get a single record from the DHT. Namely:
+
+            During a data redistribution, more than one node may be responsible
+            for a given key. In this case, the node that was most recently
+            reported as being responsible for the key is queried first, followed
+            by others (in order) until the record is located, an error occurs,
+            or no node has the record.
+
+        TODO: test the logic for retrying the request on other nodes which
+        previously covered the hash. This will require a full neo implementation
+        of the Redistribute request. See
+        https://github.com/sociomantic/dhtnode/issues/624
+
+        Params:
+            h = hash to query
+            node_hash_ranges = buffer to receive hash range information of
+                nodes which cover the specified hash
+            use_node = delegate to call to gain access to a request-on-conn to
+                communicate with the selected node
+            get = delegate to communicate with the selected node (over the
+                request-on-conn provided by use_node). Should return true if the
+                get succeeded or false to try the next node
+
+        Returns:
+            true if there are no nodes covering the specified hash
+
+    ***************************************************************************/
+
+    public bool getFromNode ( hash_t h, ref NodeHashRange[] node_hash_ranges,
+        UseNodeDg use_node,
+        bool delegate ( RequestOnConn.EventDispatcher ) get )
+    {
+        auto nodes = this.getNodesForHash(h, node_hash_ranges);
+        if ( nodes.length == 0 )
+            return true;
+
+        foreach_reverse ( node_hash_range; nodes )
+        {
+            bool try_next_node;
+            use_node(node_hash_range.addr,
+                ( RequestOnConn.EventDispatcher conn )
+                {
+                    try_next_node = get(conn);
+                }
+            );
+
+            // If we got the record or an error occurred, don't try more nodes
+            if ( !try_next_node )
+                break;
+        }
+
+        return false;
```

Comment:
This fixed `return false` doesn't seem to be covering the case where we get the list of nodes responsible for the key, but they all return `true`. Perhaps rewrite this as:

```
+            // If we got the record or an error occurred, don't try more nodes
 +            if ( !try_next_node )
 +                return true;
 +        }
 +
 +        return false;
```

Reply:
The caller can already detect this case (they are the ones who decide whether on the return value of get), so it doesn't need to be specifically covered here.

Reply:
Actually, the return value isn't needed at all. The caller can detect all possible cases externally.